service/src/BatchServer.cpp

```diff
@@ -0,0 +1,250 @@
+                    break;
+            }
+            m_batch_status->send_message(M_MESSAGE_CONTINUE);
+        }
+    }
+
+    bool BatchServerImp::is_active(void) const
+    {
+        return m_is_active;
+    }
+
+    void BatchServerImp::push_requests(void)
+    {
+        for (const auto &req : m_signal_config) {
+            m_signal_handle.push_back(
+                m_pio.push_signal(req.name, req.domain, req.domain_idx));
+        }
+        for (const auto &req : m_control_config) {
+            m_control_handle.push_back(
+                m_pio.push_control(req.name, req.domain, req.domain_idx));
+        }
+    }
+
+    void BatchServerImp::read_and_update(void)
+    {
+        if (m_signal_config.size() == 0) {
+            return;
+        }
+
+        m_pio.read_batch();
+        double *shmem_buffer = (double *)m_signal_shmem->pointer();
+        int buffer_idx = 0;
+        for (const auto &handle : m_signal_handle) {
+            shmem_buffer[buffer_idx] = m_pio.sample(handle);
+            ++buffer_idx;
+        }
+    }
+
+    void BatchServerImp::update_and_write(void)
+    {
+        if (m_control_config.size() == 0) {
+            return;
+        }
+
+        double *shmem_buffer = (double *)m_control_shmem->pointer();
+        int buffer_idx = 0;
+        for (const auto &handle : m_control_handle) {
+            m_pio.adjust(handle, shmem_buffer[buffer_idx]);
+            ++buffer_idx;
+        }
+        m_pio.write_batch();
+    }
+
+
+    void BatchServerImp::create_shmem(void)
+    {
+        // Create shared memory regions
+        size_t signal_size = m_signal_config.size() * sizeof(double);
+        size_t control_size = m_control_config.size() * sizeof(double);
+        std::string shmem_prefix_signal = M_SHMEM_PREFIX + m_server_key + "-signal";
+        std::string shmem_prefix_control = M_SHMEM_PREFIX + m_server_key + "-control";
+        int uid = pid_to_uid(m_client_pid);
+        int gid = pid_to_gid(m_client_pid);
+        if (signal_size != 0) {
+            m_signal_shmem = SharedMemory::make_unique_owner_secure(
+                shmem_prefix_signal, signal_size);
+            // Requires a chown if server is different user than client
+            m_signal_shmem->chown(uid, gid);
+        }
+        if (control_size != 0) {
+            m_control_shmem = SharedMemory::make_unique_owner_secure(
+                shmem_prefix_control, control_size);
+            // Requires a chown if server is different user than client
+            m_control_shmem->chown(uid, gid);
+        }
+    }
+
+    int BatchServerImp::fork_with_setup(std::function<void(void)> setup,
+                                        std::function<void(void)> run)
+    {
+        int pipe_fd[2];
+        check_return(pipe(pipe_fd), "pipe(2)");
+        int forked_pid = fork();
+        check_return(forked_pid, "fork(2)");
+        if (forked_pid == 0) {
+            check_return(close(pipe_fd[0]), "close(2)");
```

Comment:
Looks like the reason for this pipe is to ensure that the parent process does not proceed until the child process has created the shared memory. Why not just create the shared memory before forking?

Reply:
A good question.  The answer is that we do not want the lifetime of the shared memory regions to be coupled to the lifetime of the geopmd systemd service process (this should be since server boot).  If we create them after we fork, they should be cleaned up when the forked server ends.

Reply:
I have reset the shared memory pointers and am now calling _Exit().  This should clean up any stale keys that could get left around in certain error conditions, but does not call the destructor for all of the IOGroups registered in PlatformIO :)